src/QirRuntime/lib/Tracer/tracer.cpp

```diff
@@ -0,0 +1,305 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <cassert>
+#include <set>
+#include <sstream>
+
+#include "tracer.hpp"
+
+using namespace std;
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    thread_local std::shared_ptr<CTracer> tracer = nullptr;

```

Comment:
I would even add `static` (if that works) such that nobody else can directly access `tracer` outside of this .cpp (even those inside of `Microsoft::Quantum` namespace).
The `static` will also help to avoid the identifier conflict if someone outside of this .cpp (but still inside of `Microsoft::Quantum` namespace) adds the identifier `tracer` for a different purpose.

Reply:
The `tracer` is also accessed from tracer-qis.cpp